### PR TITLE
PIL-1295: electionUKGAAP field validation

### PR DESCRIPTION
**Pull request by MoDweik**
This PR makes it possible to validate _UKTRSubmissionData_ using the retrieved Subscription and _plrReference_.

### CODE REVIEW

File: app/uk/gov/hmrc/pillar2externalteststub/controllers/SubscriptionController.scala [hunk 1]
```diff
@@ -34,25 +32,9 @@ class SubscriptionController @Inject() (
...
-        case "XEPLR0123456500" =>
-          Future.successful(InternalServerError(Json.toJson(ServerError500.response)))
-        case "XEPLR0123456503" =>
-          Future.successful(ServiceUnavailable(Json.toJson(ServiceUnavailable503.response)))
-        case "XEPLR5555555555" =>
-          Future.successful(Ok(Json.toJson(SubscriptionSuccessResponse.successfulDomesticOnlyResponse(plrReference))))
-        case "XEPLR1234567890" =>
-          Future.successful(Ok(Json.toJson(SubscriptionSuccessResponse.successfulNonDomesticResponse(plrReference))))
-        case "XEPLR0987654321" =>
-          Future.successful(Ok(Json.toJson(NilReturnSuccess.successfulResponse)))
```
Comment (MoDweik) [inline app/uk/gov/hmrc/pillar2externalteststub/controllers/SubscriptionController.scala:53]: Also documented in the README, I'm unsure why a NilReturnSuccess is being returned from this controller and not SubmitUKTRController, isn't the Nil Return a subtype of UKTR?

Screenshot 2024-12-15 at 19 43 31

Comment (MoDweik) [inline app/uk/gov/hmrc/pillar2externalteststub/controllers/SubscriptionController.scala:53]: Also documented in the README, I'm unsure why a NilReturnSuccess is being returned from this controller and not SubmitUKTRController, isn't the Nil Return a subtype of UKTR?

The file seems to belong to neither package now.

Screenshot 2024-12-15 at 19 46 22

File: app/uk/gov/hmrc/pillar2externalteststub/models/NilReturnSuccess.scala [hunk 2]
```diff
@@ -17,15 +17,16 @@
 package uk.gov.hmrc.pillar2externalteststub.models
 
 import play.api.libs.json.{Json, OFormat}
-import uk.gov.hmrc.pillar2externalteststub.models.uktr.UktrSubmission
+import uk.gov.hmrc.pillar2externalteststub.models.subscription.SubscriptionResponse
+import uk.gov.hmrc.pillar2externalteststub.models.uktr.UKTRSubmission
 
-case class NilReturnSuccess(processingDate: String, message: String)
+case class NilReturnSuccess(processingDate: String, message: String) extends SubscriptionResponse
```
Comment (MoDweik) [inline app/uk/gov/hmrc/pillar2externalteststub/models/NilReturnSuccess.scala:23]: This builds on the presumption that NilReturnSuccess is a Subscription response as has been followed prior to this PR.

File: app/uk/gov/hmrc/pillar2externalteststub/controllers/SubscriptionController.scala (hunk 1 above)
Comment (stipoa) [reply app/uk/gov/hmrc/pillar2externalteststub/controllers/SubscriptionController.scala:53]: Yes, a good spot. 
It looks like this case class should belong in the response package :  
/pillar2-external-test-stub/app/uk/gov/hmrc/pillar2externalteststub/models/uktr/response
I can refactor this as part of : PIL-1297 Validation of MNE/UK UKTR Nil Return [which I am working on finishing today].

File: app/uk/gov/hmrc/pillar2externalteststub/models/NilReturnSuccess.scala (hunk 2 above)
Comment (stipoa) [reply app/uk/gov/hmrc/pillar2externalteststub/models/NilReturnSuccess.scala:23]: Yes, that makes sense - as per the case statement in object SubscriptionResponse - this is the response returned when a user asks for the details for a previosuly submitted Nil Return.

**Review by stipoa — APPROVED** [review]
Looks Good.

File: app/uk/gov/hmrc/pillar2externalteststub/controllers/SubmitUKTRController.scala [hunk 3]
```diff
@@ -43,7 +42,8 @@ class SubmitUKTRController @Inject() (
     with Logging {
 
   def submitUKTR: Action[JsValue] = (Action andThen authFilter).async(parse.json) { implicit request =>
-    request.headers.get("X-Pillar2-Id") match {
+    val plrReference = request.headers.get("X-Pillar2-Id")
```
Comment (ridow-hmrc) [inline app/uk/gov/hmrc/pillar2externalteststub/controllers/SubmitUKTRController.scala:45]: this doesn't seem right. Can you double check please. We should just match on Some(x) and None

File: app/uk/gov/hmrc/pillar2externalteststub/helpers/SubscriptionHelper.scala [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+object SubscriptionHelper {
+  def retrieveSubscription(plrReference: String): (Status, SubscriptionResponse) =
+    plrReference match {
+      case "XEPLR0123456404" => (NotFound, NotFoundSubscription.response)
+      case "XEPLR0123456500" => (InternalServerError, ServerError500.response)
+      case "XEPLR0123456503" => (ServiceUnavailable, ServiceUnavailable503.response)
+      case "XEPLR5555555555" => (Ok, successfulDomesticOnlyResponse(plrReference))
+      case "XEPLR1234567890" => (Ok, successfulNonDomesticResponse(plrReference))
+      case "XEPLR0987654321" => (Ok, NilReturnSuccess.successfulResponse)
```
Comment (ridow-hmrc) [inline app/uk/gov/hmrc/pillar2externalteststub/helpers/SubscriptionHelper.scala:32]: there are two concepts being mixed here. A NilReturn success seems to be for UKTR submit, whereas this is for subscriptions

Comment (MoDweik) [reply app/uk/gov/hmrc/pillar2externalteststub/helpers/SubscriptionHelper.scala:32]: I don't understand why a NilReturnSuccess would be returned from the subscription controller. What do you suggest?

File: README.md [hunk 5]
```diff
@@ -57,27 +57,26 @@ non-domestic status and specific error responses.
 
 Response Codes and Conditions:
 
-| plrReference                 | HTTP Status               | Description                                                                    |
-|:-----------------------------|---------------------------|--------------------------------------------------------------------------------|
-| XEPLR0123456404              | 404 Not Found             | Subscription not found                                                         |
```
Comment (MoDweik) [inline README.md:62]: redundant (duplicates the wildcard case)

File: app/uk/gov/hmrc/pillar2externalteststub/helpers/SubscriptionHelper.scala [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+import play.api.mvc.Results._
+import uk.gov.hmrc.pillar2externalteststub.models.subscription.SubscriptionSuccessResponse.{successfulDomesticOnlyResponse, successfulNonDomesticResponse}
+import uk.gov.hmrc.pillar2externalteststub.models.subscription._
+
+object SubscriptionHelper {
+  def isDomesticOnly(plrReference: String): Boolean =
+    retrieveSubscription(plrReference)._2 match {
+      case success: SubscriptionSuccessResponse => success.upeDetails.domesticOnly
+      case _ => throw new IllegalStateException("Expected SubscriptionSuccessResponse")
```
Comment (MoDweik) [inline app/uk/gov/hmrc/pillar2externalteststub/helpers/SubscriptionHelper.scala:27]: any thoughts on what can be returned in L27?

**Review by ridow-hmrc — APPROVED** [review]

**Review by ridow-hmrc — APPROVED** [review]

**Comment by ridow-hmrc** [comment]
I have done testing on this PR and attached evidence to each jira

**Pull request merged**